### ISSUE TITLE
Java 9 support

### ISSUE THREAD

**Issue by d1231** [HotswapProjects/HotswapAgent#218]
Hello,

What the current status for java 9 support?

Thanks

**Comment by skybber** [comment]
HA is prepared for java9, but DCEVM for java9 has still some problems with some kind of redefinition. We're working on it - it is not so simple like in case of HotswapAgent.

**Comment by d1231** [comment]
Thanks for your answer

**Comment by skybber** [comment]
Since build system for dcevm7 and dcevm8 is not able to use for java9, we had to extend dcevm from standard OpnJDK and we've created a new repository for the project as well. Jdk-hotswap9 (dcevm9) beta version is just available at [HotswapProject](https://github.com/HotswapProjects/openjdk-jdk9/releases/tag/jdk9-b94-dcevm-beta). It is bundled with HotswapAgent by default.

**Comment by boonen** [comment]
@skybber The supplied downloads for Java 9 only contain Windows builds. Do you know when Linux builds will be available?

**Comment by skybber** [comment]
I can build it for Linux as well, it is not hard do do it at all. But our ultimate intention is to distribute DCEVM builds under AdoptOpenJDK umbrella in the near future.

**Comment by skybber** [comment]
Dcevm9 **prerelease** for Linux is [here](https://github.com/HotswapProjects/openjdk-jdk9/releases/tag/jdk9-b94-dcevm-beta). For completeness, I have wrote description how to build it. Keep in mind that there are probably bugs that can cause unpredicted JVM crashes. As I know, there is problem with com.sun.proxy redefinition at least. Definitively we need help from community with testing now !

**Comment by boonen** [comment]
@skybber Thanks for putting this online so swift! I'm going to test the hotswap agent and help you improving it.

**Comment by skybber** [comment]
Hi we have no MacOS build now. It shouldn't be difficult since the build was improved in j9. Dcevm build should be part of [AdoptOpenJDK](https://github.com/AdoptOpenJDK/openjdk-build) in the future.

**Comment by theigl** [comment]
@skybber: Thanks for the quick response! Can you elaborate a bit? Does this mean that DCEVM will be packaged with the default builds of AdoptOpenJDK or will there be a custom DCEVM build? Do you have any idea when JDK10 builds will be available?

**Comment by skybber** [comment]
- Intention is to make OpenJDK with advanced hotswap support, no need to specify `-XXaltjvm=dcevm` with integrated hotswap-agent core. we want to use OpenJDK infastructure to build it.
- JDK10 for win64+linux are already available. MacOS is problem, since nobody of us have Mac. Any help is welcomed...

**Comment by theigl** [comment]
@skybber: I will try to build JDK10 from https://github.com/HotswapProjects/openjdk-jdk10 on my machine. If it works, I'll send you the artifacts and build instructions.

**Comment by theigl** [comment]
I managed to build [JDK10 with DCEVM](https://github.com/HotswapProjects/openjdk-jdk10) locally on MacOS 10.13.5:

Build Steps:
```
bash configure \
    --with-version-build="181" \
    --enable-unlimited-crypto \
    --disable-warnings-as-errors \
    --disable-hotspot-gtest

make images
```

Result:
```
./build/macosx-x86_64-normal-server-release/images/jdk/bin/java -version
openjdk version "10-internal" 2018-03-20
OpenJDK Runtime Environment (build 10-internal+181-adhoc.thomas.openjdk-jdk10)
OpenJDK 64-Bit Server VM (build 10-internal+181-adhoc.thomas.openjdk-jdk10, mixed mode)
```

When I try to instrument my project, I get the following errors:

```
OTSWAP AGENT: 13:16:07.039 ERROR (org.hotswap.agent.util.HotswapTransformer) - Error transforming class 'ch/qos/logback/core/joran/GenericConfigurator'.
java.lang.ExceptionInInitializerError
	at org.hotswap.agent.config.PluginConfiguration.initExtraClassPath(PluginConfiguration.java:178)
	at org.hotswap.agent.config.PluginConfiguration.init(PluginConfiguration.java:159)
	at org.hotswap.agent.config.PluginConfiguration.<init>(PluginConfiguration.java:62)
	at org.hotswap.agent.config.PluginManager.initClassLoader(PluginManager.java:171)
	at org.hotswap.agent.annotation.handler.PluginClassFileTransformer.transform(PluginClassFileTransformer.java:146)
	at org.hotswap.agent.annotation.handler.PluginClassFileTransformer.transform(PluginClassFileTransformer.java:89)
	at org.hotswap.agent.util.HotswapTransformer.transform(HotswapTransformer.java:184)
	at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:246)
	at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:560)
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1009)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:545)
	at java.base/java.net.URLClassLoader.access$100(URLClassLoader.java:83)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:453)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:447)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:446)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:566)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:499)
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1009)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:545)
	at java.base/java.net.URLClassLoader.access$100(URLClassLoader.java:83)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:453)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:447)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:446)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:566)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:499)
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1009)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:545)
	at java.base/java.net.URLClassLoader.access$100(URLClassLoader.java:83)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:453)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:447)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:446)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:566)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:499)
	at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredConstructors(Class.java:3090)
	at java.base/java.lang.Class.getConstructor0(Class.java:3295)
	at java.base/java.lang.Class.getConstructor(Class.java:2110)
	at org.springframework.boot.logging.LoggingSystem.get(LoggingSystem.java:169)
	at org.springframework.boot.logging.LoggingSystem.lambda$get$1(LoggingSystem.java:161)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet.lambda$entryConsumer$0(Collections.java:1571)
	at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1812)
	at java.base/java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$UnmodifiableEntrySetSpliterator.tryAdvance(Collections.java:1590)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at org.springframework.boot.logging.LoggingSystem.get(LoggingSystem.java:161)
	at org.springframework.boot.context.logging.LoggingApplicationListener.onApplicationStartingEvent(LoggingApplicationListener.java:190)
	at org.springframework.boot.context.logging.LoggingApplicationListener.onApplicationEvent(LoggingApplicationListener.java:170)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:127)
	at org.springframework.boot.context.event.EventPublishingRunListener.starting(EventPublishingRunListener.java:68)
	at org.springframework.boot.SpringApplicationRunListeners.starting(SpringApplicationRunListeners.java:48)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:313)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:137)
	at com.synexit.friendsic.view.FriendseekApplication.main(FriendseekApplication.java:58)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.springframework.boot.maven.AbstractRunMojo$LaunchRunner.run(AbstractRunMojo.java:496)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.NullPointerException
	at org.hotswap.agent.javassist.util.proxy.SecurityActions.setAccessible(SecurityActions.java:103)
	at org.hotswap.agent.javassist.util.proxy.DefineClassHelper.toClass3(DefineClassHelper.java:153)
	at org.hotswap.agent.javassist.util.proxy.DefineClassHelper.toClass2(DefineClassHelper.java:136)
	at org.hotswap.agent.javassist.util.proxy.DefineClassHelper.toClass(DefineClassHelper.java:95)
	at org.hotswap.agent.javassist.util.proxy.FactoryHelper.toClass(FactoryHelper.java:133)
	at org.hotswap.agent.javassist.util.proxy.ProxyFactory.createClass3(ProxyFactory.java:530)
	at org.hotswap.agent.javassist.util.proxy.ProxyFactory.createClass2(ProxyFactory.java:515)
	at org.hotswap.agent.javassist.util.proxy.ProxyFactory.createClass1(ProxyFactory.java:451)
	at org.hotswap.agent.javassist.util.proxy.ProxyFactory.createClass(ProxyFactory.java:422)
	at org.hotswap.agent.util.classloader.URLClassLoaderHelper.<clinit>(URLClassLoaderHelper.java:47)
	... 79 more
```

Then followed by hundreds of exceptions like this:

```
OTSWAP AGENT: 13:16:23.903 ERROR (org.hotswap.agent.util.HotswapTransformer) - Error transforming class 'org/springframework/cglib/core/EmitUtils$6'.
java.lang.NoClassDefFoundError: Could not initialize class org.hotswap.agent.util.classloader.URLClassLoaderHelper
	at org.hotswap.agent.config.PluginConfiguration.initExtraClassPath(PluginConfiguration.java:178)
	at org.hotswap.agent.config.PluginConfiguration.init(PluginConfiguration.java:159)
	at org.hotswap.agent.config.PluginConfiguration.<init>(PluginConfiguration.java:62)
	at org.hotswap.agent.config.PluginManager.initClassLoader(PluginManager.java:171)
	at org.hotswap.agent.annotation.handler.PluginClassFileTransformer.transform(PluginClassFileTransformer.java:146)
	at org.hotswap.agent.annotation.handler.PluginClassFileTransformer.transform(PluginClassFileTransformer.java:89)
	at org.hotswap.agent.util.HotswapTransformer.transform(HotswapTransformer.java:184)
	at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:246)
	at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:560)
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1009)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:174)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:545)
	at java.base/java.net.URLClassLoader.access$100(URLClassLoader.java:83)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:453)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:447)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:446)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:566)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:499)
	at org.springframework.cglib.core.EmitUtils.string_switch_hash(EmitUtils.java:270)
	at org.springframework.cglib.core.EmitUtils.string_switch(EmitUtils.java:172)
	at org.springframework.cglib.proxy.MethodInterceptorGenerator.generateFindProxy(MethodInterceptorGenerator.java:232)
	at org.springframework.cglib.proxy.MethodInterceptorGenerator.generate(MethodInterceptorGenerator.java:133)
	at org.springframework.cglib.proxy.Enhancer.emitMethods(Enhancer.java:1219)
	at org.springframework.cglib.proxy.Enhancer.generateClass(Enhancer.java:630)
	at org.springframework.cglib.transform.TransformingClassGenerator.generateClass(TransformingClassGenerator.java:33)
	at org.springframework.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanFactoryAwareGeneratorStrategy.generate(ConfigurationClassEnhancer.java:253)
	at org.springframework.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:329)
	at org.springframework.cglib.proxy.Enhancer.generate(Enhancer.java:492)
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:93)
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:91)
	at org.springframework.cglib.core.internal.LoadingCache$2.call(LoadingCache.java:54)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at org.springframework.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:61)
	at org.springframework.cglib.core.internal.LoadingCache.get(LoadingCache.java:34)
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData.get(AbstractClassGenerator.java:116)
	at org.springframework.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:291)
	at org.springframework.cglib.proxy.Enhancer.createHelper(Enhancer.java:480)
	at org.springframework.cglib.proxy.Enhancer.createClass(Enhancer.java:337)
	at org.springframework.context.annotation.ConfigurationClassEnhancer.createClass(ConfigurationClassEnhancer.java:138)
	at org.springframework.context.annotation.ConfigurationClassEnhancer.enhance(ConfigurationClassEnhancer.java:110)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.enhanceConfigurationClasses(ConfigurationClassPostProcessor.java:403)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanFactory(ConfigurationClassPostProcessor.java:254)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:284)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:128)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:694)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:532)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:140)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:395)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:327)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:137)
```

Should the most recent agent version 1.3.0 work with JDK10? Do I need another agent version or some different configuration for JDK10?

**Comment by skybber** [comment]
Could you try to disable ProxyPlugin?

**Comment by theigl** [comment]
My app starts with this minimal configuration. I had to remove `extraClassPath`.

```
autoHotswap=true
disabledPlugins=Proxy,Hibernate,Spring,Tomcat
LOGGER=info
```

Startup is extremely slow because the Plugin seems to be re-initialized all the time.

```
[INFO] --- spring-boot-maven-plugin:2.0.3.RELEASE:run (default-cli) ---
HOTSWAP AGENT: 14:55:15.057 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'java.net.URLClassLoader@7b91d9f'.
HOTSWAP AGENT: 14:55:16.287 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'lombok.launch.ShadowClassLoader@1d1c37d5'.
HOTSWAP AGENT: 14:55:17.551 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'java.net.URLClassLoader@3a749d7a'.
HOTSWAP AGENT: 14:55:19.921 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@7e6c717'.
HOTSWAP AGENT: 14:55:20.197 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@37918b3f'.
HOTSWAP AGENT: 14:55:20.389 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@7581274c'.
HOTSWAP AGENT: 14:55:20.612 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@4ea117fc'.
HOTSWAP AGENT: 14:55:20.852 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@7f72662'.
HOTSWAP AGENT: 14:55:21.799 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.logback.LogbackPlugin' initialized in ClassLoader 'java.net.URLClassLoader@3a749d7a'.
HOTSWAP AGENT: 14:55:24.518 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@75f459c7'.
HOTSWAP AGENT: 14:55:24.766 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@7bf0dd72'.
HOTSWAP AGENT: 14:55:24.856 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@2abaaa07'.
HOTSWAP AGENT: 14:55:24.974 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@13eaa3e0'.
HOTSWAP AGENT: 14:55:25.163 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@becaac3'.
HOTSWAP AGENT: 14:55:25.294 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@4bb66917'.
HOTSWAP AGENT: 14:55:25.393 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@6f4cd8da'.
HOTSWAP AGENT: 14:55:25.633 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@3c3e4bc9'.
HOTSWAP AGENT: 14:55:25.719 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@304a3c4d'.
HOTSWAP AGENT: 14:55:25.806 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@2c08f366'.
HOTSWAP AGENT: 14:55:25.911 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@4bf61812'.
HOTSWAP AGENT: 14:55:26.087 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@42d3ece0'.
HOTSWAP AGENT: 14:55:26.178 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@165f7323'.
HOTSWAP AGENT: 14:55:26.280 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@4fe939fd'.
HOTSWAP AGENT: 14:55:26.383 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@6bfe30e0'.
HOTSWAP AGENT: 14:55:26.476 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@4135be9b'.
HOTSWAP AGENT: 14:55:26.630 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@1ed543b3'.
HOTSWAP AGENT: 14:55:26.725 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@6e728649'.
```

**Comment by skybber** [comment]
Could you create minimalistic application demonstrating this problem?

**Comment by theigl** [comment]
I can try this with a Spring Boot "Hello World" app tomorrow.

**Comment by theigl** [comment]
The exact same thing happens with the default spring boot template app.

Go to https://start.spring.io, and generate a project with dependency "Web".

Add the following hotswap-agent.properties:

```
autoHotswap=true
LOGGER=info
```

And run it with `mvn spring-boot:run`.

Output sample:

```
HOTSWAP AGENT: 18:06:20.608 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@68c0477f'.
HOTSWAP AGENT: 18:06:20.688 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@192a7550'.
HOTSWAP AGENT: 18:06:20.778 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@44e1aed6'.
HOTSWAP AGENT: 18:06:20.886 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@57c7066d'.
HOTSWAP AGENT: 18:06:20.965 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@3d28afe5'.
HOTSWAP AGENT: 18:06:21.064 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@66d5372b'.
HOTSWAP AGENT: 18:06:21.144 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@1b72ab50'.
HOTSWAP AGENT: 18:06:21.257 INFO (org.hotswap.agent.config.PluginRegistry) - Plugin 'org.hotswap.agent.plugin.hotswapper.HotswapperPlugin' initialized in ClassLoader 'jdk.internal.reflect.DelegatingClassLoader@532b766a'.
```

If I add anything to `extraClasspath` startup fails with the same exceptions I posted above.

**Comment by theigl** [comment]
Reloading classes works, but also generates dozens of log messages for a single reload:

```
HOTSWAP AGENT: 18:10:51.718 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.718 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.718 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.719 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.719 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.719 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.719 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.719 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.719 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
HOTSWAP AGENT: 18:10:51.719 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.example.demo.DemoApplication] (autoHotswap)
```

**Comment by theigl** [comment]
I think HotswapAgent needs to exclude `jdk.internal.reflect.DelegatingClassLoader`, called `sun.reflect.DelegatingClassLoader` until JDK9. 

Is there a configuration options to ignore certain classloaders so I could give this a try?

**Comment by skybber** [comment]
Shuld be fixed in 1.3.1.SNAPSHOT + dcevm11
